src/test/fuzz/util.cpp

```diff
@@ -524,7 +525,28 @@ CNetAddr ConsumeNetAddr(FuzzedDataProvider& fuzzed_data_provider) noexcept
     } else if (network == Network::NET_INTERNAL) {
         net_addr.SetInternal(fuzzed_data_provider.ConsumeBytesAsString(32));
     } else if (network == Network::NET_ONION) {
-        net_addr.SetSpecial(fuzzed_data_provider.ConsumeBytesAsString(32));
+        auto pub_key{fuzzed_data_provider.ConsumeBytes<uint8_t>(torv3::ADDR_SIZE)};
+        if (pub_key.size() > 0 && pub_key[0] & 1) {
+            pub_key.resize(32);
+            // Create valid onion address to prevent the address checksum from
+            // being a fuzz blocker.
+            uint8_t check_sum[torv3::CHECKSUM_LEN];
+            torv3::Checksum({pub_key.data(), pub_key.size()}, check_sum);
+
+            std::vector<uint8_t> onion_addr;
+            onion_addr.insert(onion_addr.end(), pub_key.begin(), pub_key.end());
+            onion_addr.push_back(check_sum[0]);
+            onion_addr.push_back(check_sum[1]);
+            onion_addr.push_back(torv3::VERSION[0]);
+            assert(onion_addr.size() == torv3::TOTAL_LEN);
+
+            auto encoded{EncodeBase32({onion_addr.data(), onion_addr.size()})};
+            encoded += ".onion";
+            net_addr.SetSpecial(encoded);
+        } else {
+            std::string addr(reinterpret_cast<const std::string::value_type*>(pub_key.data()), pub_key.size());
+            net_addr.SetSpecial(addr);
```

Comment:
`SetSpecial()` expects a string like `5g72ppm3krkorsfopcm2bi7wlv4ohhs4u4mlseymasn7g7zhdcyjpfid.onion` and this code will feed it a binary string (containing non-printable chars, maybe also `'\0'`) of length 32. It will _never_ succeed because 32 is too short.

I think that the branch `} else if (network == Network::NET_ONION) {` should always do its best to generate a valid Tor address. Maybe append `Network::NET_UNROUTABLE` to the array at the start of `ConsumeNetAddr()` and don't do anything in that case (leave `net_addr` default constructed, which is `!IsValid()`) in order to generate invalid results some of the time.

Reply:
I wanted to preserve the previous behavior but you are right that will never work in any case because of the length.

Now it only ever produces valid onion addresses. I didn't add `NET_UNROUTABLE` as i think that would invalidate all our corpuses that make use of `ConsumeNetAddr` or at least make them less useful (not sure how much we would/should care about that).